micropython/drivers/storage/sdcard/sdcard.py

```diff
@@ -63,89 +77,89 @@ def init_spi(self, baudrate):
-        if csd[0] & 0xC0 == 0x40:  # CSD version 2.0
-            self.sectors = ((csd[8] << 8 | csd[9]) + 1) * 1024
-        elif csd[0] & 0xC0 == 0x00:  # CSD version 1.0 (old, <=2GB)
-            c_size = (csd[6] & 0b11) << 10 | csd[7] << 2 | csd[8] >> 6
-            c_size_mult = (csd[9] & 0b11) << 1 | csd[10] >> 7
-            read_bl_len = csd[5] & 0b1111
+        self.CSD = csd_int = int.from_bytes(
+            csd, "big"
+        )  # convert 16-byte CSD to a giant integer for bit extraction
+        gb = self.gb  # just for shorter code
+        # use bit numbers from SD card spec v9.0.0, table 5.3.2
+        vers = gb(csd_int, 126, 127)
+        if vers == 1:  # CSD version 2.0
+            self.sectors = (gb(csd_int, 48, 69) + 1) * 1024
+            self.cdv = 1
+        elif vers == 0x00:  # CSD version 1.0 (old, <=2GB)
+            c_size = gb(csd_int, 62, 73)
+            c_size_mult = gb(csd_int, 47, 49)
+            read_bl_len = gb(csd_int, 80, 83)
             capacity = (c_size + 1) * (2 ** (c_size_mult + 2)) * (2**read_bl_len)
             self.sectors = capacity // 512
+            self.cdv = 512  # converts bytes to sectors
         else:
-            raise OSError("SD card CSD format not supported")
+            raise OSError(EIO, "CSD format unknown")
         # print('sectors', self.sectors)
 
         # CMD16: set block length to 512 bytes
-        if self.cmd(16, 512, 0) != 0:
-            raise OSError("can't set 512 block size")
+        if self.cmd(16, 512, None) != 0:
+            raise OSError(EIO, "can't set 512 block size")
 
         # set to high data rate now that it's initialised
         self.init_spi(baudrate)
 
-    def init_card_v1(self):
-        for i in range(_CMD_TIMEOUT):
-            time.sleep_ms(50)
-            self.cmd(55, 0, 0)
-            if self.cmd(41, 0, 0) == 0:
-                # SDSC card, uses byte addressing in read/write/erase commands
-                self.cdv = 512
-                # print("[SDCard] v1 card")
-                return
-        raise OSError("timeout waiting for v1 card")
-
-    def init_card_v2(self):
-        for i in range(_CMD_TIMEOUT):
-            time.sleep_ms(50)
-            self.cmd(58, 0, 0, 4)
-            self.cmd(55, 0, 0)
-            if self.cmd(41, 0x40000000, 0) == 0:
-                self.cmd(58, 0, 0, -4)  # 4-byte response, negative means keep the first byte
-                ocr = self.tokenbuf[0]  # get first byte of response, which is OCR
-                if not ocr & 0x40:
-                    # SDSC card, uses byte addressing in read/write/erase commands
-                    self.cdv = 512
-                else:
-                    # SDHC/SDXC card, uses block addressing in read/write/erase commands
-                    self.cdv = 1
-                # print("[SDCard] v2 card")
-                return
-        raise OSError("timeout waiting for v2 card")
-
     def cmd(self, cmd, arg, crc, final=0, release=True, skip1=False):
```

Comment:
I suggest removing the `crc` argument and just always computing it in this function.  The only times it's passed in as non-None is for a few initialisation/detection commands at the start, and that's only done once so doesn't need to be efficient (in time).

Removing the `crc` argument will make code size smaller and faster for the case when `None` is passed (which is the majority of uses).

Reply:
I agree with this.  The passing in of fixed CRCs is historically done for the few initialization packets, but why?  Do you want me to make the change in my fork, or do you want to do this after you pull it?

Reply:
is this OK style for the crc table, including expectation of memory footprint after parsing?:

```python
crc7_be_syndrome_table = (
b'\x00\x12$6HZl~\x90\x82\xb4\xa6\xd8\xca\xfc\xee2 \x16\x04zh^L\xa2\xb0\x86\x94\xea\xf8'
b'\xce\xdcdv@R,>\x08\x1a\xf4\xe6\xd0\xc2\xbc\xae\x98\x8aVDr`\x1e\x0c:(\xc6\xd4\xe2\xf0'
b'\x8e\x9c\xaa\xb8\xc8\xda\xec\xfe\x80\x92\xa4\xb6XJ|n\x10\x024&\xfa\xe8\xde\xcc\xb2\xa0'
b'\x96\x84jxN\\"0\x06\x14\xac\xbe\x88\x9a\xe4\xf6\xc0\xd2<.\x18\ntfPB\x9e\x8c\xba\xa8'
b'\xd6\xc4\xf2\xe0\x0e\x1c*8FTbp\x82\x90\xa6\xb4\xca\xd8\xee\xfc\x12\x006$ZH~l\xb0\xa2'
b'\x94\x86\xf8\xea\xdc\xce 2\x04\x16hzL^\xe6\xf4\xc2\xd0\xae\xbc\x8a\x98vdR@>,\x1a\x08'
b'\xd4\xc6\xf0\xe2\x9c\x8e\xb8\xaaDV`r\x0c\x1e(:JXn|\x02\x10&4\xda\xc8\xfe\xec\x92\x80'
b'\xb6\xa4xj\\N0"\x14\x06\xe8\xfa\xcc\xde\xa0\xb2\x84\x96.<\n\x18ftBP\xbe\xac\x9a\x88\xf6'
b'\xe4\xd2\xc0\x1c\x0e8*TFpb\x8c\x9e\xa8\xba\xc4\xd6\xe0\xf2'
)
```

Reply:
> The passing in of fixed CRCs is historically done for the few initialization packets, but why? Do you want me to make the change in my fork, or do you want to do this after you pull it?

I'm not exactly sure why the CRC was there only for the few initialisation packets.  But my guess would be 1) it's a static value so doesn't need a function to compute it, 2) for initialisation it may be that some cards require the CRC on these commands.

Anyway, I'm happy for you to make this change as part of this PR.



> is this OK style for the crc table, including expectation of memory footprint after parsing?

Yes that looks OK.  Maybe indent each line within the parenthesis, if `ruff format` allows?  (we switched from black to ruff).  Eg:
```py
table = (
    b'...'
    b'...'
)
```

Also I suggest putting this crc7 function within `sdcard.py`.  It's simple enough and saves having separate files for the driver, ie everything is self contained in one file.